### DISCUSSION TITLE
How to verify generic return type in BuildEligibility for async methods?

### DISCUSSION THREAD

**Discussion by joakimriedel** [metalama/Metalama#318] [Q&A] [answered]
New to metalama, so this might be obvious but cannot find any help through the docs.

I would like to check something like

```c#
builder.MustSatisfy(m => 
    m.ReturnType.Is(typeof(Task<ServiceResponse>)) || 
    m.ReturnType.Is(typeof(Task<ServiceResponse<>>)), 
    m => $"{m} must return Task<ServiceResponse> or Task<ServiceResponse<T>>");
```

but this obviously wont work since `typeof(Task<ServiceResponse<>>)` is not valid C#.

I tried the following, but it won't work due since runtime reflection cannot be used at compile time.

```c#
builder.MustSatisfy(m => 
    m.ReturnType.Is(typeof(Task<ServiceResponse>)) || 
    m.ReturnType.ToType().GenericTypeArguments[0].GetGenericTypeDefinition() == typeof(ServiceResponse<>), 
    m => $"{m} must return Task<ServiceResponse> or Task<ServiceResponse<T>>");
```

Resorted to the following solution, which works, but there must be a better way..

```c#
builder.MustSatisfy(m =>
    m.ReturnType.ToString()!.Contains("ServiceResponse"),
    m => $"{m} must return Task<ServiceResponse> or Task<ServiceResponse<T>>");
```

Please help me understand how I will properly test for a generic return type for async methods.

**Comment by svick** [comment, marked as answer]
You can use `Is` with `ConversionKind.TypeDefinition` for this:

```c#
builder.ReturnType().MustSatisfy(
    type => type.Is(typeof(Task<ServiceResponse>)) ||
        (type.Is(typeof(Task<>), ConversionKind.TypeDefinition) &&
        ((INamedType)type).TypeArguments.Single().Is(typeof(ServiceResponse<>), ConversionKind.TypeDefinition)),
    t => $"{t} must be Task<ServiceResponse> or Task<ServiceResponse<T>>");
```

(I also changed the code to use `builder.ReturnType()`, but that's not necessary.)

Another option is to use `GetAsyncInfo()`, which will work for any awaitable type (like `ValueTask<T>`):

```c#
builder.ReturnType().MustSatisfy(
    type =>
    {
        var asyncInfo = type.GetAsyncInfo();

        return asyncInfo.IsAwaitable &&
            (asyncInfo.ResultType.Is(typeof(ServiceResponse)) ||
            asyncInfo.ResultType.Is(typeof(ServiceResponse<>), ConversionKind.TypeDefinition));        
    }, t => $"{t} must be Task<ServiceResponse> or Task<ServiceResponse<T>>");
```

**Reply by joakimriedel** [reply]
Yes, this is the solution! Thanks 🙏

I actually did find, and try, using `GetAsyncInfo()` but I did not know about the `ConversionKind.TypeDefinition` flag so I tested for `(asyncInfo.ResultType.Is(typeof(ServiceResponse)) || asyncInfo.ResultType.Is(typeof(ServiceResponse<>))` which would not work. This would make an excellent note in some docs for the future!
